src/core/Akka/Util/Internal/AtomicState.cs

```diff
@@ -0,0 +1,186 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="AtomicState.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+//     Copyright (C) 2013-2015 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+
+namespace Akka.Util.Internal
+{
+    /// <summary>
+    /// Internal state abstraction
+    /// </summary>
+    internal abstract class AtomicState : AtomicCounterLong, IAtomicState
+    {
+        private readonly ConcurrentQueue<Action> _listeners;
+        private readonly TimeSpan _callTimeout;
+
+        protected AtomicState( TimeSpan callTimeout, long startingCount )
+            : base( startingCount )
+        {
+            _listeners = new ConcurrentQueue<Action>( );
+            _callTimeout = callTimeout;
+        }
+
+        /// <summary>
+        /// Add a listener function which is invoked on state entry
+        /// </summary>
+        /// <param name="listener">listener implementation</param>
+        public void AddListener( Action listener )
+        {
+            _listeners.Enqueue( listener );
+        }
+
+        /// <summary>
+        /// Test for whether listeners exist
+        /// </summary>
+        public bool HasListeners
+        {
+            get { return !_listeners.IsEmpty; }
+        }
+
+        /// <summary>
+        /// Notifies the listeners of the transition event via a 
+        /// </summary>
+        protected void NotifyTransitionListeners( )
+        {
+            if ( !HasListeners ) return;
+            Task
+                .Factory
+                .StartNew
```

Comment:
@Horusiath @rogeralsing No -- it could lead to undesirable and hard to debug side effects depending on what type of actions users pass into the "AddListener" method. The original akka.io code adds objects which extend "Runnable" to a queue. It then passes each runnable to an executor to execute without blocking.

Since I've decided to ignore ExecutionContext for now -- the best way for me to handle this is to propagate the Async through the internal / protected methods up to the "CallThrough method" and await the result there. I'll try to implement the change this way and run through the tests and verify that it doesn't block the main thread.